modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i

```diff
@@ -131,12 +134,86 @@ velocity_interp_method = 'rc'
 [Executioner]
   type = Steady
   solve_type = 'NEWTON'
-  petsc_options_iname = '-pc_type -pc_factor_shift_type'
-  petsc_options_value = 'lu NONZERO'
   nl_rel_tol = 1e-12
 []
 
+[Preconditioning]
+  active = SMP
+  [FSP]
+    type = FSP
+    # It is the starting point of splitting
+    topsplit = 'up' # 'up' should match the following block name
+    [up]
+      splitting = 'u p' # 'u' and 'p' are the names of subsolvers
+      splitting_type  = schur
+      # Splitting type is set as schur, because the pressure part of Stokes-like systems
+      # is not diagonally dominant. CAN NOT use additive, multiplicative and etc.
+      #
+      # Original system:
+      #
+      # | Auu Aup | | u | = | f_u |
+      # | Apu 0   | | p |   | f_p |
+      #
+      # is factorized into
+      #
+      # |I             0 | | Auu  0|  | I  Auu^{-1}*Aup | | u | = | f_u |
+      # |Apu*Auu^{-1}  I | | 0   -S|  | 0  I            | | p |   | f_p |
+      #
+      # where
+      #
+      # S = Apu*Auu^{-1}*Aup
+      #
+      # The preconditioning is accomplished via the following steps
+      #
+      # (1) p* = f_p - Apu*Auu^{-1}f_u,
+      # (2) p = (-S)^{-1} p*
+      # (3) u = Auu^{-1}(f_u-Aup*p)
+
+      petsc_options_iname = '-pc_fieldsplit_schur_fact_type  -pc_fieldsplit_schur_precondition -ksp_gmres_restart -ksp_rtol -ksp_type'
+      petsc_options_value = 'full                            selfp                             300                1e-4      fgmres'
+    []
+    [u]
+      vars = 'vel_x vel_y'
+      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_type -ksp_rtol -ksp_gmres_restart'
+      petsc_options_value = 'hypre    boomeramg      fgmres    5e-1      300'
```

Comment:
Is there a good reason to use FGMRES as an inner solver?

Reply:
I cannot answer you from the linear algebra perspective. As you know I didn't even know that it was invalid to be using gmres in the outer linear solve. But in my timings of my application, I see fgmres with the inner solves running faster than gmres with the inner solves.

| dofs | inner ksp solve type | time |
| --- | --- | --- |
| 384000 | fgmres | 19.392 |
| 384000 | gmres | 24.897 |
| 1536000 | fgmres | 25.598 |
| 1536000 | gmres | 35.807 |

Reply:
What about `-ksp_type gmres -ksp_pc_side right`?

Reply:
The difference you are seeing between `KSPFGMRES` versus `KSPGMRES` + `PC_RIGHT` is because the latter applies the `PC` an extra time, but it does not need to store an extra Krylov basis (which is the case of the former).

Reply:
I love getting your input on this PR. I've been learning so much!

Reply:
Flexible Krylov methods necessarily apply preconditioners on the right. You can't do `-ksp_type fgmres -ksp_pc_side left`, that will throw an error. You can do `-ksp_type fgmres -ksp_pc_side right`, but that's just redundant.